File: InterviewPreparationKit/String Manipulation/MakingAnagrams.cpp
// https://www.hackerrank.com/challenges/ctci-making-anagrams/problem
#include <bits/stdc++.h>

using namespace std;

// Complete the makeAnagram function below.
int makeAnagram(string a, string b) {
    map<char, int> a_map;
    map<char, int> b_map;

    for (int i = 0; i < a.length(); i++) {
        char letter = a[i];
        if (a_map.count(letter) > 0) {
            a_map[letter] += 1;
        }
        else {
            a_map[letter] = 1;
        }
    }

    for (int j = 0; j < b.length(); j++) {
        char letter = b[j];
        if (b_map.count(letter) > 0) {
            b_map[letter] += 1;
        }
        else {
            b_map[letter] = 1;
        }
    }

    int removed = 0;
    for (char key = 'a'; key <= 'z'; key++) {
        if (a_map.count(key) > 0 && b_map.count(key) > 0) {
            removed += abs(a_map[key] - b_map[key]);
        }
        else if (a_map.count(key) > 0) {
            removed += a_map[key];
        }
        else if (b_map.count(key) > 0) {
            removed += b_map[key];
        }
    }

    return removed;
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string a;
    getline(cin, a);

    string b;
    getline(cin, b);

    int res = makeAnagram(a, b);

    fout << res << "\n";

    fout.close();

    return 0;
}

File: Interview Preparation Kit/String Manipulation/MakingAnagrams.js
// https://www.hackerrank.com/challenges/ctci-making-anagrams/problem
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

// Complete the makeAnagram function below.
function makeAnagram(a, b) {
    let a_map = new Map();
    let b_map = new Map();

    for (let i = 0; i < a.length; i++) {
        let letter = a[i];
        if (a_map.has(letter)) {
            a_map.set(letter, a_map.get(letter) + 1);
        }
        else {
            a_map.set(letter, 1);
        }
    }

    for (let j = 0; j < b.length; j++) {
        let letter = b[j];
        if (b_map.has(letter)) {
            b_map.set(letter, b_map.get(letter) + 1);
        }
        else {
            b_map.set(letter, 1);
        }
    }

    let removed = 0;
    for (let keyIndex = 0; keyIndex < 26; keyIndex++) {
        let key = String.fromCharCode(keyIndex + 97);
        if (a_map.has(key) && b_map.has(key)) {
            removed += Math.abs(a_map.get(key) - b_map.get(key));
        }
        else if (a_map.has(key)) {
            removed += a_map.get(key);
        }
        else if (b_map.has(key)) {
            removed += b_map.get(key);
        }
    }

    return removed;
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const a = readLine();

    const b = readLine();

    const res = makeAnagram(a, b);

    ws.write(res + '\n');

    ws.end();
}
